lib/src/main/java/com/auth0/jwt/algorithms/Algorithm.java

```diff
@@ -364,9 +364,11 @@ public String toString() {
     /**
      * Sign the given content using this Algorithm instance.
      *
-     * @param contentBytes an array of bytes representing the base64 encoded content to be verified against the signature.
+     * @param headerBytes an array of bytes representing the base64 encoded header content to be verified against the signature.
+     * @param payloadBytes an array of bytes representing the base64 encoded payload content to be verified against the signature.
      * @return the signature in a base64 encoded array of bytes
      * @throws SignatureGenerationException if the Key is invalid.
      */
-    public abstract byte[] sign(byte[] contentBytes) throws SignatureGenerationException;
+    public abstract byte[] sign(byte[] headerBytes, byte[] payloadBytes) throws SignatureGenerationException;
```

Comment:
Yes. Basically this narrows the use for Algorithm to just signing tokens, which is in line with its own class description (in the javadoc).

Reply:
I don't disagree that this is an improvement. But it will break customers who are implementing their own Algorithms.. I can't see many customers doing that but some feature requests have been opened in the past asking us to extend the algorithms we support. Let me discuss this internally. Maybe with a good readme note and proper changelog notes we can accept it. I can't promise anything.

Reply:
Re above: No breaking change allowed (and no major plans for the time being). We need to find an alternative to this. Normally I'd say call the new method from the old one and deprecate it, but in this case there's no way for you to figure out at what point of the `contentBytes` array is the half of it in order to separate it into header and payload bytes. What if you add the new sign signature and keep both methods side-to-side? I guess the new one can call the old one by merging both header and payload bytes with a "." in-between, so not to repeat the sign method's body code. If not you will have to rollback the changes for this class.

Reply:
I've added a default implementation of the new method (sign with header and payload) in Algorithm, but overridden that in the known subclasses. This should ensure backwards compatibility for those implementing their own algorithms, at the same time as speeding things up for the rest of us. 

CryptoHelper has now both the new and the old methods, but they're all quite small.

Reply:
@lbalmaceda If this solution is acceptable, I'll add a few of the old unit tests for code coverage (see failing build checks).